snapshot/snapshot_computation.py

```diff
@@ -0,0 +1,305 @@
+
+Detailed documentation: https://precice.org/tooling-micro-manager-overview.html
+"""
+
+import argparse
+import importlib
+import os
+import sys
+from mpi4py import MPI
+import numpy as np
+import logging
+import time
+
+from .config_snapshot import SnapshotConfig
+from .read_write_hdf import ReadWriteHDF as hp
+from micro_manager.micro_simulation import create_simulation_class
+
+sys.path.append(os.getcwd())
+
+
+class SnapshotComputation:
+    def __init__(self, config_file: str) -> None:
+        """
+        Constructor.
+
+        Parameters
+        ----------
+        config_file : string
+            Name of the JSON configuration file (provided by the user).
+        """
+        self._comm = MPI.COMM_WORLD
+        self._rank = self._comm.Get_rank()
+        self._size = self._comm.Get_size()
+
+        self._logger = logging.getLogger(__name__)
+        self._logger.setLevel(level=logging.INFO)
+
+        # Create file handler which logs messages
+        fh = logging.FileHandler("snapshot-computation.log")
+        fh.setLevel(logging.INFO)
+
+        # Create formatter and add it to handlers
+        formatter = logging.Formatter(
+            "[" + str(self._rank) + "] %(name)s -  %(levelname)s - %(message)s"
+        )
+        fh.setFormatter(formatter)
+        self._logger.addHandler(fh)  # add the handlers to the logger
+
+        self._is_parallel = self._size > 1
+        self._micro_sims_have_output = False
+
+        self._logger.info("Provided configuration file: {}".format(config_file))
+        self._config = SnapshotConfig(self._logger, config_file)
+
+        # Data names of data to output to the snapshot database
+        self._write_data_names = self._config.get_write_data_names()
+
+        # Data names of data to read as input parameter to the simulations
+        self._read_data_names = self._config.get_read_data_names()
+
+        # Path to the parameter file containing input parameters for micro simulations
+        self._parameter_file = self._config.get_parameter_file_name()
+
+        self._is_postprocessing_required = self._config.get_postprocessing()
+
+        self._merge_output_files = self._config.get_merge_output()
+
+        self._is_micro_solve_time_required = self._config.write_micro_solve_time()
+
+        self._local_number_of_sims = 0
+        self._global_number_of_sims = 0
+        self._is_rank_empty = False
+
+        self._initialize()
+
+    # **************
+    # Public methods
+    # **************
+
+    def solve(self) -> None:
+        """
+        Solve the problem by iterating over a set macro parameters.
+        - Read macro parameters from the config file, solve micro simulation one by one,
+          and write outputs to a file.
+        """
+
+        # Loop over all macro parameters
+        for elems in range(self._local_number_of_sims):
+            # Create micro simulation object
+            self._micro_sims = create_simulation_class(self._micro_problem)(
```

Comment:
In contrast to the Micro Manager, micro simulations are newly initialized in the `solve()` method as each microsimulation is only required for one loop of the `solve()` method. Is it a good solution to initalize microsimulations within the solve call as this is technically part of the `_initialize()`-method's responsibility?

Reply:
We need to be careful with the terms here. The step here is the **creation** of the micro simulations, and not their initialization. Initialization is essentially optional, and only happens if the micro simulation library has a function called `initialize()`. Initialization of the micro simulations is handled in the `initialize()` function of the MicroManager class. Inheriting from this class and calling this function would make sense.